README.md
# `vscode-extension-imaginary` README

 VSCode extension allowing to create a thumbnail from a selected URL (through a remote `Imaginary` HTTP call), and to replace this selection with the corresponding markdown tag.

 `Imaginary` is a HTTP micro-service allowing to execute some operations on images through HTTP requests (like "resize this image to 650 pixels wide and convert to PNG"), @see https://github.com/h2non/imaginary.

## Features

![](https://github.com/SR-G/vscode-extension-imaginary/raw/master/vscode-extension-imaginary.gif)

## Requirements

You need a working `Imaginary` server deployed somewhere (same host, or another server (local or remote)).

May for example be started through `docker` with  :

```bash
docker run -d -p 9000:9000 --name imaginary h2non/imaginary -cors -gzip -enable-url-source -enable-auth-forwarding
```

## Extension Settings

This extension has to be configured, either on a global level or on a workspace level (recommended - will allow one different Imaginary usage per project).

- `baseurl` is the full HTTP url of the running `Imaginary` server (for example, "http://127.0.0.1:9000/" if `Imaginary` is running on localhost).
- `operation` is the main operation that will be triggered. 
- `parameters` is a map for the various sub-parameters of this operation and have to be entered accordingly, @see [Imaginary documentation](https://github.com/h2non/imaginary#http-api).
- `destination` is the path where the converted image will be stored, under the project root folder. Some tokens will be automatically replaced, for the now : `${year}`, `${month}`, `${day}` (may be useful for blogging systems)
- `markdown-path` is the path that will be used in the markdown tag. May be different than destination (for example, for `hugo` blogs, statis fields will be stored in the `static/` sub-folder, but this level does not appear once the blog is published)

```json
// Place your settings in this file to overwrite default and user settings.
{
    "imaginary.baseurl" : "http://192.168.8.4:9000",
    "imaginary.operation" : "thumbnail",
    "imaginary.parameters" : {
        "format" : "png",
        "width" : "200",
    },
    "imaginary.destination" : "static/media/activity/${year}/${month}/",  
    "imaginary.markdown-path" : "/media/activity/${year}/${month}/",  
}
```

## Developement

**Packaging** : @see https://code.visualstudio.com/docs/extensions/publish-extension

```bash
// first steps
npm install -g vsce
vsce ls
```

```bash
// publication
npm install
vsce publish
```

**Testing the dev version** :

Generate the extension through `vsce package`, then install it either through `CTRL + SHIFT + P` > `Extensions : install from VSIX` or through `/c/Tools/vscode/1.19.0/code --install-extension vscode-extension-imaginary-0.0.1.vsix`

While developing, if needed, in order to retrieve `node_modules` under the installed folder, use :

```bash
cd  ~/.vscode/extensions/SR-G.vscode-extension-imaginary-0.0.1
npm install
```

CHANGELOG.md
# Change Log
All notable changes to the "vscode-extension-imaginary" extension will be documented in this file.

Check [Keep a Changelog](http://keepachangelog.com/) for recommendations on how to structure this file.

## 0.0.5
- Minor correction about directories creation when using a network UNC path

## 0.0.4
- Minor corrections about tokens not correctly padded with zeros

## 0.0.1
- Initial release


extension.js
// The module 'vscode' contains the VS Code extensibility API
// Import the module and reference it with the alias vscode in your code below
const vscode = require('vscode');
const path = require('path')
const fs = require('fs')
const stream = require('stream')
const request = require('request')

const padZero = function (s, len, c){
    var c = c || '0';
    while(s.toString().length < len) {
        s = c + s;
    }
    return s;
}

const createDirectories = function (targetDir) {
    var sep = "/"; // path.sep
    if (!fs.existsSync(targetDir)){      
        const initDir = path.isAbsolute(targetDir) ? sep : '';
        targetDir.split(sep).reduce((parentDir, childDir) => {
        const curDir = path.resolve(parentDir, childDir);
        if (!fs.existsSync(curDir)){      
            fs.mkdirSync(curDir);
        }
        return curDir;
        }, initDir);    
    }
}

const replaceTokens = function(s) {
    const now = new Date();
    const day = now.getDate();
    const month = now.getMonth() + 1;
    const year = now.getFullYear();
    return s.replace('\${month}', padZero(month, 2, "0")).replace('\${year}', year).replace('\${day}', padZero(day, 2, "0"));
}

const sanitizePath = function(s) {
    return s.replace(/\\/g, "/").replace(/(?!^)\/\//g, "/");
}

// this method is called when your extension is activated
// your extension is activated the very first time the command is executed
function activate(context) {

    // Use the console to output diagnostic information (console.log) and errors (console.error)
    // This line of code will only be executed once when your extension is activated
    console.log('Congratulations, your extension "vscode-extension-imaginary" is now active!');


    // The command has been defined in the package.json file
    // Now provide the implementation of the command with  registerCommand
    // The commandId parameter must match the command field in package.json
    let commandRenameMarkdownImage = vscode.commands.registerCommand('extension.renameMarkdownImage', function () {
        const workspace0 = vscode.workspace.workspaceFolders ? vscode.workspace.workspaceFolders[0] : undefined;
        const folderName = workspace0 ? workspace0.uri.fsPath : undefined;

        var editor = vscode.window.activeTextEditor;
        if (!editor) {
            return; // No open text editor
        }

        // selected text, will be used during the HTTP call
        var selection = editor.selection;
        var text = editor.document.getText(selection);
        const configuration = vscode.workspace.getConfiguration("imaginary"); 
        var destination = configuration.get('destination', '');
        var markdownPath = configuration.get('markdown-path', destination);
        destination = replaceTokens(destination);
        markdownPath = replaceTokens(markdownPath);
        var currentBasename = path.dirname(text);
        var currentFileName = path.basename(text);

        vscode.window.showInputBox({prompt: 'What should be the new name ?', value: currentFileName}).then(function(val) {
            var currentLocation = sanitizePath(`${folderName}/${destination}/${currentFileName}`);
            var newLocation = sanitizePath(`${folderName}/${destination}/${val}`);
            if (!fs.existsSync(currentLocation)){
                currentLocation = sanitizePath(`${folderName}/${markdownPath}/${currentFileName}`);
                newLocation = sanitizePath(`${folderName}/${markdownPath}/${val}`);
                if (!fs.existsSync(currentLocation)){
                    vscode.window.showErrorMessage("Can't find current image anywhere : " + text + '');  
                    return;
                }
            }

            if (fs.existsSync(newLocation)) {
                vscode.window.showErrorMessage('New file already found on disk : ' + val + '');  
                return;
            }

            fs.rename(currentLocation, newLocation, function(err) {
                if ( err ) { 
                    console.log('ERROR: ' + err);
                    vscode.window.showErrorMessage('Error on renaming to : ' + val + '');  
                } else {
                    editor.edit(function (builder) {
                        builder.replace(selection, text.replace(currentFileName, val));
                    });                    
                }
            });            
        });
    });

    let commandImaginaryOperation = vscode.commands.registerCommand('extension.imaginaryOperation', function () {
        const workspace0 = vscode.workspace.workspaceFolders ? vscode.workspace.workspaceFolders[0] : undefined;
        const folderName = workspace0 ? workspace0.uri.fsPath : undefined;
        
        var editor = vscode.window.activeTextEditor;
        if (!editor) {
            return; // No open text editor
        }

        // selected text, will be used during the HTTP call
        var selection = editor.selection;
        var text = editor.document.getText(selection);
        
        // read configuration
        const configuration = vscode.workspace.getConfiguration("imaginary"); 
        const baseUrl = configuration.get("baseurl", "http://127.0.0.1:9000");
        const action = configuration.get("operation", "thumbnail");
        const extension = configuration.get('format', 'png');
        const height = configuration.get('size', '200');
        var destination = configuration.get('destination', '');
        var markdownPath = configuration.get('markdown-path', destination)

        // allow destination to have tokens like ${month}, ${year}, ${day} (usefull for blogging systems)
        destination = replaceTokens(destination);
        markdownPath = replaceTokens(markdownPath);

        // query parameters
        var parameters = configuration.get('parameters', {});
        var sep = "?";
        var qs = "";
        for (var key in parameters) {
            if (parameters.hasOwnProperty(key)) {
                qs = qs + sep + key + "=" + parameters[key];
            }
            sep = "&";
        }
        qs = qs + sep + "url=" + text

        // final URL, image name, directory where image will be saved
        const url = `${baseUrl}/${action}${qs}`;
        var imageName =  text.split("/").pop();
        imageName = imageName.substr(0, imageName.lastIndexOf(".")) + "." + extension;
        var dir = `${folderName}/${destination}`
        dir = sanitizePath(dir);
        createDirectories(dir);
        var dest = `${folderName}/${destination}/${imageName}`;
        dest = sanitizePath(dest);

        // execute the request + image saving and notify user if sucessful / error
        var stream = request(url).on('error', function(err) {
            console.log(err)
            vscode.window.showErrorMessage('Error on conversion from url : ' + url + '');  
          }).on('response', function(response) {
            console.log(response.statusCode) // 200
            console.log(response.headers['content-type']) // 'image/png'
          }).pipe(fs.createWriteStream(dest));

        stream.on('finish', function () { 
            vscode.window.showInformationMessage('Image successfully written to : ' + dest + '');  
        }).on('error', function (err) {
            vscode.window.showErrorMessage('Error while writing image to : ' + dest + '');  
        });

        // replace the initialy selected text with the markdown image tag containing the path to the saved image
        // todo : would be nice to have a customized file name, but would need to have a second parameter somewhere (in addition to the selected URL)
        editor.edit(function (builder) {
            var s = markdownPath
            if (s == "") {
                s = "/";
            }
            builder.replace(selection, "![](" + s + imageName + ")");
        });
    });

    context.subscriptions.push(commandImaginaryOperation);
    context.subscriptions.push(commandRenameMarkdownImage);
}
exports.activate = activate;

// this method is called when your extension is deactivated
function deactivate() {
}
exports.deactivate = deactivate;